Wall flow interface kernel diode

GiudGiud opened the pull request:
Early PR for feedback

I think this may be needed fairly soon.
I ll test it more. I have encouraging results in that test though not 100% satisfactory either

I m thinking we should rework the force boundary execution system a little more than this and only keep three parameters:
force_all_boundary_execution
boundaries_to_force (always active, not just with boolean)
boundaries_to_avoid (same)

what do you think @lindsayad

GiudGiud committed:
Add option to avoid internal boundaries for flux kernels
refs #20695

GiudGiud committed:
Use option to skip internal sidesets in INSFVAdvection
refs #20695

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1008821/) on ca3ddda wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20743/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20743/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 1a35e4abdee91fdc1cf4f887b7e4934827a64255`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1010320/) on 8d0eda3 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20743/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20743/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 1a35e4abdee91fdc1cf4f887b7e4934827a64255`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1010369/) on 6ac275f wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/20743/site)

This comment will be updated on new commits.

GiudGiud committed:
Add tests for rework of flux kernel boundary forcing/skippingm refs #20695

GiudGiud committed:
Suppress forcing parameters from INSFV advection, add boundary skipping, closes #20742

GiudGiud committed:
Make boundary skipping/forcing attributes private to force using an API, refs #20742

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1010805/) on b8dc26b wanted to post the following:

The following commit(s) contain fixups:

- 8a306c71d3: fixup! Add option to avoid internal boundaries for flux kernels refs #20695

GiudGiud committed:
Remove duplicate parameter in INSFV advection kernel refs #20772

lindsayad commented:
> force_all_boundary_execution boundaries_to_force (always active, not just with boolean) boundaries_to_avoid (same)
> 
> what do you think @lindsayad

sounds fine to me

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1010904/) on 5da099c wanted to post the following:

The following commit(s) contain fixups:

- 8a306c71d3: fixup! Add option to avoid internal boundaries for flux kernels refs #20695

GiudGiud committed:
Fixup test for friction flow diode, refs #20695

GiudGiud commented:
This should be ready for review @lindsayad

GiudGiud committed:
Dont use nodal values on FV variables, refs #15836

lindsayad commented on framework/src/fvkernels/FVFluxKernel.C:
```diff
@@ -86,27 +88,23 @@ FVFluxKernel::onBoundary(const FaceInfo & fi) const
     if (fi.boundaryIDs().find(bnd_to_force) != fi.boundaryIDs().end())
       return false;
 
+  // We're not on a boundary, so in practice we're not 'skipping'
+  if (!onBoundary(fi))
+    return false;
```
So you want to be able to avoid internal boundaries using this new logic above (e.g. the `avoidBoundary` method)? Otherwise I would keep this up top because it is the most common case. Otherwise you are doing O(N) loops over the boundaries to force and O(log N) set finds in the inner loop for no reason

GiudGiud replied:
Yeah I want to be able to skip internal boundaries. I ll move this a little higher, to skip the search over boundaries_to_force

lindsayad commented on modules/navier_stokes/src/fviks/INSFVMassWallFlowDiode.C:
```diff
@@ -0,0 +1,59 @@
+
+  auto v = _rc_vel_provider.getVelocity(_velocity_interp_method, *_face_info, _tid);
+
+  if (_inverted * v * _normal < 0)
+    v = 0;
```
Maybe you should just early return 0 here and avoid the following somewhat-expensive interpolation

GiudGiud commented:
I tagged on a few random fixes to help some ongoing modeling

GiudGiud committed:
Address Alex's review

GiudGiud committed:
Adapt test to new restriction on outputting an element norm to auxvariable, refs #15836

GiudGiud committed:
Add material with friction term turning on/off based on velocity
refs  #20695

GiudGiud committed:
Remove friction diode kernel, will just make it a material, refs #20695

GiudGiud committed:
Use a functor for the characteristic length in PINSFV, refs #21065

GiudGiud committed:
Modify the friction diode material that adds a friction term to the existing one, refs #20695
- allow for quadratic term
- have a boolean to turn it on and off instead of relying on an element per element criterion

GiudGiud committed:
Add test for the friction diode material, refs #20695

GiudGiud committed:
Add a piecewise by block vector functor material for purpose of adding friction anisotropic coefficients, refs #20695

GiudGiud committed:
Fix documentation for piecewise functor material
Add docs for piecewise vector functor material, refs  #20695

GiudGiud commented:
@lindsayad this should be ready for a round of review.

note that the boundaries_to_avoid feature is not used currently. @grmnptr will find it useful for his work on refactoring the backend I think, and it ll be used in the future wall diode which requires a one-sided gradient (which we do have but not inside the domain)

lindsayad commented on framework/include/fvkernels/FVFluxKernel.h:
```diff
@@ -124,10 +124,17 @@ class FVFluxKernel : public FVKernel,
+   * Returns whether to avoid execution on a boundary
+   * @param fi the FaceInformation currently considered
+   */
+  bool avoidBoundary(const FaceInfo & fi) const
+  {
+    for (const auto bnd_id : fi.boundaryIDs())
+      if (_boundaries_to_avoid.count(bnd_id))
+        return true;
+    return false;
```
put this in .C or make it `inline` outside of the class definition

GiudGiud replied:
moved in .C

lindsayad commented on modules/navier_stokes/include/materials/NSFVFrictionFlowDiodeMaterial.h:
```diff
@@ -0,0 +1,42 @@
+
+private:
+  /// Direction of the diode
+  const RealVectorValue & _direction;
+
+  /// Magnitude of the additional linear resistance
+  const RealVectorValue & _linear_resistance;
+
+  /// Magnitude of the additional quadratic resistance
+  const RealVectorValue & _quadratic_resistance;
```
```suggestion
  const RealVectorValue _direction;

  /// Magnitude of the additional linear resistance
  const RealVectorValue _linear_resistance;

  /// Magnitude of the additional quadratic resistance
  const RealVectorValue _quadratic_resistance;
```
Avoids querying the pointer to the memory location (not that this is likely to be important)

lindsayad commented on modules/navier_stokes/include/materials/NSFVFrictionFlowDiodeMaterial.h:
```diff
@@ -0,0 +1,42 @@
+  const Moose::Functor<ADRealVectorValue> & _base_quadratic_friction;
+
+  /// Whether the diode is active or not
+  const bool & _diode_on;
```
```suggestion
  const bool _diode_on;
```

GiudGiud replied:
needed if it s to be controllable

lindsayad reviewed (changes requested):
overall looks good to me

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1068581/) on fd597cd wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>8888f0</th>
      <th colspan=4>#20743 fd597c</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.56%</td>
      <td>82.57%</td>
      <td>+0.01%</td>
      <td>96.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>75533</td>
      <td>75541</td>
      <td>+8</td>
      <td>24</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15955</td>
      <td>15948</td>
      <td>-7</td>
      <td>1</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20743/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20743/coverage/framework)


## Modules coverage

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>8888f0</th>
      <th colspan=4>#20743 fd597c</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.55%</td>
      <td>85.59%</td>
      <td>+0.04%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>9697</td>
      <td>9723</td>
      <td>+26</td>
      <td>47</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1638</td>
      <td>1637</td>
      <td>-1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20743/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20743/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/20743/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/20743/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/20743/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/20743/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/20743/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/20743/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/20743/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/20743/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/20743/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/20743/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/20743/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/20743/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/20743/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/20743/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/20743/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/20743/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/20743/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/20743/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/20743/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/20743/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/20743/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/20743/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/20743/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/20743/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud committed:
Add documentation to friction flow diode material, refs #20695
